Change objcImpl syntax to `@objc @implementation`

beccadax opened the pull request:
This PR changes the type-checking rules for objcImpl to require an `@objc` attribute. It also moves the category name from the `@implementation` attribute to the `@objc` attribute. Backwards compatibility is maintained for the pre-stabilization `@_objcImplementation` syntax.

As part of this change, plain `@objc(CustomName)` on an `extension` now changes its category name when it is printed through `PrintAsClang` or IRGen'd; the name previously was parsed and partially validated but ignored. The only new diagnostic related to this is a warning in Swift 5 mode but will be an error in Swift 6 mode.

~~This draft currently incorporates commits from #73128 and will need to be rebased once that lands, so don't worry about the gigantic Reviewers list it has right now.~~

beccadax committed:
[NFC] Add `@objc` to imported ObjC categories

This contains the category name, if there is one.

beccadax committed:
Test stable and unstable objcImpl separately

Their syntaxes are about to diverge, so let’s make sure that we maintain source compatibility for @_objcImplementation.

beccadax committed:
Require @objc to be used with @implementation

…for extensions. This change also removes @implementation(CategoryName); you should attach the category name to the @objc attribute instead. And there are small changes to how much checking the compiler will do on an @objc @implementation after the decl checker has discovered a problem with it.

beccadax committed:
[NFC] Merge two category name accessors

beccadax committed:
Support @objc(CustomName) on extensions

This now specifies a category name that’s used in TBDGen, IRGen, and PrintAsClang. There are also now category name conflict diagnostics; these subsume some @implementation diagnostics.

(It turns out there was already a check for @objc(CustomName) to make sure it wasn’t a selector!)

beccadax commented:
@swift-ci please test

beccadax committed:
Fix objcImpl resilience tests

beccadax commented:
@swift-ci please test

beccadax committed:
Change expected objcImpl private let optimization

Previous changes have made it so that `private let`s in an objcImpl extension are implicitly `@objc`, which changed downstream behavior in a SILOptimizer test. The new behavior is actually more correct, so codify this behavior change by modifying the test.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci Please Build Toolchain macOS Platform

beccadax commented:
@swift-ci Please Build Toolchain Linux Platform

beccadax commented:
Won't be able to build a toolchain until apple/llvm-project#8714 merges.

nkcsgexi commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -3614,6 +3812,18 @@ class ObjCImplementationChecker {
     }
   }
 
+  static SmallString<64> makeObjCAttrInsertionText(Identifier categoryName) {
```
Nitpick: is there a way we could share the two copies of `makeObjCAttrInsertionText`?

beccadax replied:
Yeah, I was being a little lazy here. Let me see if I can clean this up.

nkcsgexi replied:
Thank you!

beccadax replied:
Took some doing, but I've generalized `fixDeclarationObjCName()` so I can apply it to extensions instead of writing a one-off helper for objcImpl to use.

beccadax commented:
@swift-ci please build toolchain

beccadax commented:
Toolchains:

* macOS: https://ci.swift.org/job/swift-PR-toolchain-macos/1247/artifact/branch-main/swift-PR-73309-1247-osx.tar.gz
* Ubuntu: https://download.swift.org/tmp/pull-request/73309/820/ubuntu2004/PR-ubuntu2004.tar.gz
* Windows: https://ci-external.swift.org/job/swift-PR-build-toolchain-windows/1235/artifact/build/artifacts/

(Of course, this is only really going to work on Apple platforms, since it’s an ObjC interop feature.)

beccadax commented on lib/AST/NameLookup.cpp:
```diff
@@ -2262,6 +2267,27 @@ void NominalTypeDecl::recordObjCMethod(AbstractFunctionDecl *method,
+    return;
+
+  // Use a linear scan on the assumption that there aren't very many extensions.
+  for (auto *other : getExtensions()) {
```
I probably don't want the map to be per-`ClassDecl` because the vast majority of classes won't have *any* extensions with a category name, but maybe it can be per-module or something, with a `std::pair<ClassDecl *, Identifier>` key.

tshortli replied:
Request result storage goes in a `DenseMap` unless you choose to customize it to be inline, and you'd only need to allocate the extension map result for classes that have `recordObjCCategory()` called on them so in practice I think it wouldn't create much overhead to do it per-class.

beccadax replied:
The problem is that adding an extension invalidates the lookup table, and `recordCategory(ext)` is called immediately after `ext` is added to its class, so it's likely the table will be incomplete. But if we defer the conflict-checking logic until after we've added all the extensions to the class, we can then use a lookup table to accelerate conflict checking. It's just a bigger redesign of this logic than "drop a lookup table into this existing function", which is what I thought you were suggesting!

tshortli replied:
I see, that makes sense. If you were to delay this check to after all extensions were registered, no extra storage would be needed I guess because it can just be a single pass over the complete list of extensions.

beccadax replied:
Done. While I was working on it, I discovered that some of my request caching code was horribly busted, so fixing that was also nice. :)

beccadax committed:
[NFC] Fix objcImpl request cache

Several offsetting bugs both broke the caching of `ObjCInterfaceAndImplementationRequest` and caused it to usually miss. Fix this whole painful mess. Also has collateral improvements to simple_display().

beccadax committed:
[NFC] Hoist getAttributeInsertionLoc() up to Decl

This functionality was previously reserved for ValueDecls. Move it all the way up to Decl; in the process, make it correctly handle EnumElementDecls and EnumCaseDecls.

This change also allows us to generalize `swift::fixDeclarationObjCName()` to work on extensions, though we do not use that capability in this commit.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test
